asv_bench/benchmarks/groupby.py

```diff
@@ -91,313 +114,259 @@ def time_groupby_first(self, dtype):
-        self.N = 10000
-        self.labels = np.random.randint(0, 2000, size=self.N)
-        self.labels2 = np.random.randint(0, 3, size=self.N)
-        self.df = DataFrame({
-            'key': self.labels,
-            'key2': self.labels2,
-            'value1': np.random.randn(self.N),
-            'value2': (['foo', 'bar', 'baz', 'qux'] * (self.N // 4)),
-        })
-
-    @staticmethod
-    def scalar_function(g):
-        return 1
+    def setup(self, ncols):
+        N = 1000
+        data = np.random.randn(N, ncols)
+        self.labels = np.random.randint(0, 100, size=N)
+        self.df = DataFrame(data)
 
-    def time_groupby_frame_apply_scalar_function(self):
-        self.df.groupby(['key', 'key2']).apply(self.scalar_function)
+    def time_sum(self, ncols):
+        self.df.groupby(self.labels).sum()
 
-    def time_groupby_frame_apply_scalar_function_overhead(self):
-        self.df.groupby('key').apply(self.scalar_function)
 
-    @staticmethod
-    def df_copy_function(g):
-        # ensure that the group name is available (see GH #15062)
-        g.name
-        return g.copy()
+class Nth(object):
 
-    def time_groupby_frame_df_copy_function(self):
-        self.df.groupby(['key', 'key2']).apply(self.df_copy_function)
-
-    def time_groupby_frame_apply_df_copy_overhead(self):
-        self.df.groupby('key').apply(self.df_copy_function)
+    goal_time = 0.2
 
+    def setup_cache(self):
+        df = DataFrame(np.random.randint(1, 100, (10000, 2)))
+        df.iloc[1, 1] = np.nan
+        return df
 
-#----------------------------------------------------------------------
-# 2d grouping, aggregate many columns
+    def time_frame_nth_any(self, df):
+        df.groupby(0).nth(0, dropna='any')
 
-class groupby_frame_cython_many_columns(object):
-    goal_time = 0.2
+    def time_frame_nth(self, df):
+        df.groupby(0).nth(0)
 
-    def setup(self):
-        self.labels = np.random.randint(0, 100, size=1000)
-        self.df = DataFrame(randn(1000, 1000))
+    def time_series_nth_any(self, df):
+        df[1].groupby(df[0]).nth(0, dropna='any')
 
-    def time_sum(self):
-        self.df.groupby(self.labels).sum()
+    def time_series_nth(self, df):
+        df[1].groupby(df[0]).nth(0)
 
 
-#----------------------------------------------------------------------
-# single key, long, integer key
+class Incidies(object):
 
-class groupby_frame_singlekey_integer(object):
     goal_time = 0.2
 
     def setup(self):
-        self.data = np.random.randn(100000, 1)
-        self.labels = np.random.randint(0, 1000, size=100000)
-        self.df = DataFrame(self.data)
+        rng = date_range('1/1/2000', '12/31/2005', freq='H')
+        self.year, self.month, self.day = rng.year, rng.month, rng.day
+        self.ts = Series(np.random.randn(len(rng)), index=rng)
 
-    def time_sum(self):
-        self.df.groupby(self.labels).sum()
+    def time_datetime_indicies(self):
+        self.ts.groupby([self.year, self.month, self.day])
```

Comment:
before this was doing `len(..)` of the object I think? That is an important difference :-) (as the above is lazy, doesn't do any grouping yet)